src/js/actions/layereffects.js

```diff
@@ -325,6 +333,89 @@ define(function (require, exports) {
+                effects: source.dropShadows
+            }];
+
+        // For each effectType
+        //      For each effect at index i
+        //          Build every layer's ID array, with index for that layer, and props for that layer's new effect
+
+        // Build a different master props/index list for each effect type
+        sourceEffects.forEach(function (sourceEffect) {
+            var curType = sourceEffect.type,
+                effectPropsList = [],
+                effectIndexList = [];
+
+            masterEffectTypes.push(curType);
+            // Each effect will have either an index, or null to be inserted into each layer,
+            // which are saved in effectPropsList and effectIndexList
+            sourceEffect.effects.forEach(function (effectObj, effectIndex) {
+                var perEffectIndexList = [],
+                    perEffectPropsList = [];
+
+                targetLayers.forEach(function (targetLayer) {
+                    var targetLayerEffects = targetLayer.getLayerEffectsByType(curType),
+                        newEffectProps = effectObj;
+
+                    if (targetLayerEffects && targetLayerEffects.has(effectIndex)) {
+                        // If it already exists, we can just push the ID for that effect
+                        perEffectIndexList.push(effectIndex); // The order we push on these lists is the layer ordering
+                    } else {
+                        // If it doesn't exist, we push a null object/null index
+                        // THIS IS GONNA BE A PROBLEM WITH MULTIPLES
+                        perEffectIndexList.push(null); // null signifies this will be the first effect 
+                    }
+
+                    var enabled = (newEffectProps.enabled === undefined) || newEffectProps.enabled;
+
+                    // We have to keep this immutable so we don't lose the data structures like Color in the Record
+                    newEffectProps.set("enabled", enabled);
+
+                    perEffectPropsList.push(newEffectProps);
+                });
+                effectIndexList.push(Immutable.List(perEffectIndexList));
+                effectPropsList.push(Immutable.List(perEffectPropsList));
+            });
+            // Now that we've build per effect per layer lists, we can add them to master list
+            masterEffectIndexMap[curType] = Immutable.List(effectIndexList);
+            masterEffectPropsMap[curType] = Immutable.List(effectPropsList);
+        });
+
+        var payload = {
+            documentID: document.id,
+            layerIDs: layerIDs,
+            layerEffectTypes: Immutable.List(masterEffectTypes),
+            layerEffectIndex: Immutable.Map(masterEffectIndexMap),
+            layerEffectProps: Immutable.Map(masterEffectPropsMap),
+            coalesce: false
+        };
+
+        // Synchronously update the stores
+        this.dispatch(events.document.history.optimistic.LAYER_EFFECTS_MASS_CHANGED, payload);
```

Comment:
This is the way we do effects. We update our own stores first, so later on when we call _syncStoreToPs, we get the overall effect data from the store and pass it to Photoshop.